plugin/pkg/shoot/quotavalidator/admission.go

```diff
@@ -83,15 +123,454 @@ func (h *RejectShootIfQuotaExceeded) Admit(a admission.Attributes) error {
+		}
+		for _, metric := range quotaMetricNames {
+			allocatedResources[metric] = sumQuantity(allocatedResources[metric], shootResources[metric])
+		}
+	}
+
+	// TODO: We have to determine and add the amount of storage, which is allocated by manually created persistent volumes
+	// and the count of loadbalancer, which are created due to manually created services of type loadbalancer
+
+	return allocatedResources, nil
+}
+
+func (h *RejectShootIfQuotaExceeded) findShootsReferQuota(quota garden.Quota, shoot garden.Shoot) ([]garden.Shoot, error) {
+	var shootsReferQuota []garden.Shoot
+
+	privateSecretBindings, err := h.getPrivateSecretBindingsReferQuota(quota)
+	if err != nil {
+		return nil, err
+	}
+	crossSecretBindings, err := h.getCrossSecretBindingsReferQuota(quota)
+	if err != nil {
+		return nil, err
+	}
+
+	// Find all shoots which are referencing found PrivateSecretBindings
+	for _, binding := range privateSecretBindings {
+		shoots, err := h.shootLister.
+			Shoots(binding.Namespace).
+			List(labels.Everything())
+		if err != nil {
+			return nil, err
+		}
+		for _, s := range shoots {
+			// exclude actual shoot from resource allocation calculation (update case)
+			if shoot.Namespace == s.Namespace && shoot.Name == s.Name {
+				continue
+			}
+			if s.Spec.Cloud.SecretBindingRef.Kind == "PrivateSecretBinding" && s.Spec.Cloud.SecretBindingRef.Name == binding.Name {
+				shootsReferQuota = append(shootsReferQuota, *s)
+			}
+		}
+	}
+
+	// Find all shoots which are referencing found CrossSecretBindings
+	for _, binding := range crossSecretBindings {
+		shoots, err := h.shootLister.
+			Shoots(binding.Namespace).
+			List(labels.Everything())
+		if err != nil {
+			return nil, err
+		}
+		for _, s := range shoots {
+			// exclude actual shoot from resource allocation calculation (update case)
+			if shoot.Namespace == s.Namespace && shoot.Name == s.Name {
+				continue
+			}
+			if s.Spec.Cloud.SecretBindingRef.Kind == "CrossSecretBinding" && s.Spec.Cloud.SecretBindingRef.Name == binding.Name {
+				shootsReferQuota = append(shootsReferQuota, *s)
+			}
+		}
+	}
+	return shootsReferQuota, nil
+}
+
+func (h *RejectShootIfQuotaExceeded) getPrivateSecretBindingsReferQuota(quota garden.Quota) ([]garden.PrivateSecretBinding, error) {
+	var privateSBsReferQuota []garden.PrivateSecretBinding
+
+	privateSBs, err := h.privateSBLister.
+		PrivateSecretBindings(v1.NamespaceAll).
+		List(labels.Everything())
+	if err != nil {
+		return nil, err
+	}
+
+	for _, binding := range privateSBs {
+		for _, quotaRef := range binding.Quotas {
+			if quota.Name == quotaRef.Name && quota.Namespace == quotaRef.Namespace {
+				privateSBsReferQuota = append(privateSBsReferQuota, *binding)
+			}
+		}
+	}
+	return privateSBsReferQuota, nil
+}
+
+func (h *RejectShootIfQuotaExceeded) getCrossSecretBindingsReferQuota(quota garden.Quota) ([]garden.CrossSecretBinding, error) {
+	var crossSBsReferQuota []garden.CrossSecretBinding
+
+	crossSBs, err := h.crossSBLister.
```

Comment:
Can we write these statements into one line (line breakings are not used anywhere else in the code)?

Reply:
done